sdk/eventhubs/azure_messaging_eventhubs/examples/eventhubs_tracing_subscriber.rs

```diff
@@ -0,0 +1,217 @@
+//! and sends them to Event Hubs asynchronously.
+
+use azure_identity::DefaultAzureCredential;
+use azure_messaging_eventhubs::{models::EventData, ProducerClient};
+use std::error::Error;
+use std::sync::Arc;
+use std::{env, time::SystemTime};
+use tokio::sync::mpsc;
+use tracing::field::Visit;
+use tracing::{Event, Level, Subscriber};
+use tracing_subscriber::{
+    fmt,
+    layer::{Context, Layer},
+    prelude::*,
+    registry::{LookupSpan, Registry},
+};
+
+/// A custom tracing layer that sends log messages to Azure Event Hubs.
+struct EventHubsLayer {
+    sender: mpsc::Sender<EventData>,
+}
+
+impl EventHubsLayer {
+    async fn new(
+        fully_qualified_namespace: &str,
+        eventhub_name: &str,
+    ) -> Result<Self, Box<dyn Error>> {
+        let producer = ProducerClient::builder()
+            .open(
+                fully_qualified_namespace,
+                eventhub_name,
+                DefaultAzureCredential::new()?,
+            )
+            .await?;
+
+        let (sender, mut receiver) = mpsc::channel::<EventData>(100);
+        let producer_arc = Arc::new(producer);
+        let producer_clone = producer_arc.clone();
+
+        // Spawn a task that receives log messages and sends them to EventHubs
+        tokio::spawn(async move {
+            while let Some(event) = receiver.recv().await {
+                println!("Sending event to EventHubs: {:?}", event);
+                if let Err(err) = producer_clone.send_event(event, None).await {
+                    eprintln!("Failed to send event to EventHubs: {err}");
+                }
+            }
+        });
+
+        Ok(Self { sender })
+    }
+}
+
+/// Constants for property names in the EventData.
+/// These constants are used to set properties in the EventData object.
+const LOG_LEVEL_PROPERTY: &str = "log_level";
+const NAME_PROPERTY: &str = "name";
+const TARGET_PROPERTY: &str = "target";
+const MODULE_PATH_PROPERTY: &str = "module_path";
+const FILE_PROPERTY: &str = "file";
+const LINE_PROPERTY: &str = "line";
+const FIELDS_PROPERTY: &str = "fields";
+const IS_SPAN_PROPERTY: &str = "is_span";
+const IS_EVENT_PROPERTY: &str = "is_event";
+const TIMESTAMP_PROPERTY: &str = "timestamp";
+
+impl<S> Layer<S> for EventHubsLayer
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn on_event(&self, event: &Event<'_>, _ctx: Context<'_, S>) {
+        // Format the event into a string
+        let mut visitor = EventVisitor {
+            buffer: String::new(),
+        };
+
+        event.record(&mut visitor);
+
+        // Add timestamp, level, and target
+        let metadata = event.metadata();
+
+        // Build an EventData object containing both the event body and properties reflecting the event's metadata.
+        let mut event_data_builder = EventData::builder()
+            .with_body(visitor.buffer)
+            .add_property(LOG_LEVEL_PROPERTY.to_string(), metadata.level().as_str())
```

Comment:
Seeing all these `to_string()` calls does make me cringe a bit for the name. Makes me wonder if maybe we should use a `Cow<'static, str>` for field names. Wouldn't the names often be static strings anyway?

Reply:
We consume the strings in the API and our API guidelines say that consumed strings should be passed as `String` and not as `&str`.

Reply:
I get that and I know that's a discussion point we need to have again, but we've already made some stipulations to the guidelines (and I've updated them accordingly) to make the DX better. This is another case. If we expect that most devs are going to be coding property names (just like we do for headers, which we do ultimately accept as cows), using a `Cow<'static, str>` through the function and the structure - right up until you call `into()` to convert from your structure to fe2o3 (in which case the data gets copied & owned either way) - might be a better DX.

This isn't a blocker for this PR, though; but, I think we should have a tracking issue on it. Seeing it in use like this looks pretty gnarly.